### PR TITLE
Document deploying ASO in multitenant mode

### PR DESCRIPTION
**Pull request by babbageclunk**
**What this PR does / why we need it**:
Explains the process of deploying multiple instances of ASO in one cluster with different Azure credentials.

**Special notes for your reviewer**:
At the moment I'm splitting the YAML up manually - I don't think it would be difficult to script but I'm not sure of the best way to do it. 

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains documentation
- [ ] this PR contains tests

### CODE REVIEW

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Deploying ASO v2 in multi-tenant mode
+
+The current release version of the ASO deployment YAML installs the operator in single-tenant mode:
+it's deployed in the `azureserviceoperator-system` namespace with one set of Azure credentials, and manages resources in any namespace in the cluster.
```
Comment (matthchr) [inline docs/hugo/content/introduction/multitenant-deployment.md:4]: Should this be like a bulleted list or something? It almost reads like one.

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+That single operator deployment handles webhooks fired when Azure resources are changed.
+
+ASO can also be deployed in a multi-tenant configuration that enables using separate credentials to manage resources in different Kubernetes namespaces.
+
+Running the operator in multi-tenant mode requires one deployment to handle webhooks (since webhook configurations are cluster-level resources) and then a deployment for each tenant, each with its own credentials and set of namespaces that it watches for Azure resources.
+To deploy the operator in multi-tenant mode the release YAML has been split into two parts:
+
+1. Cluster-wide resources:
+   * Custom resource definitions for the Azure resources
```
Comment (matthchr) [inline docs/hugo/content/introduction/multitenant-deployment.md:13]: minor: consistency - you end some of these with a period and some without. I think they should all end in `.`

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+   * Cluster roles for managing those resources
+   * The `azureserviceoperator-system` namespace containing the deployment and service to run the operator.
```
Comment (matthchr) [inline docs/hugo/content/introduction/multitenant-deployment.md:15]: A bit confused about this. In multitenant mode this namespace would just be for the webhook deployment, since it's running in webhook only mode, right?

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+     The webhook deployment of the operator is configured to run in webhook-only mode.
```
Comment (matthchr) [inline docs/hugo/content/introduction/multitenant-deployment.md:16]:
```suggestion
     The webhook deployment if the operator is configured to run in webhook-only mode.
```

I think?

Comment (matthchr) [reply docs/hugo/content/introduction/multitenant-deployment.md:16]: Combine with the above comment though I think it might be clearer to just talk about this as the webhooks for the multi-tenant operator (or something) though, rather than calling it "the operator" which confused me a bit because there's kinda more than one of them.

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+## Example files
+Examples of the deployment YAML files are available on the release page for ASO v2 releases from [v2.0.0-alpha.5](https://github.com/Azure/azure-service-operator/releases/tag/v2.0.0-alpha.5).
+The cluster-wide file `multitenant-cluster_v2.0.0-alpha.5.yaml` can be used as-is (the webhook operator namespace is fixed as `azureserviceoperator-system`),
+but the namespaces and cluster role binding in the per-tenant file `multitenant-tenant_v2.0.0-alpha.5.yaml` will need to be updated from `tenant1` to the desired name for each tenant.
+
+## Per-tenant configuration
+Create the `aso-controller-settings` secret as described in the [authentication docs](https://azure.github.io/azure-service-operator/introduction/authentication/),
+but create the secret in the tenant namespace and add an extra target namespaces key to it:
+```
+export TENANT_NAMESPACE="<tenant namespace>"
```
Comment (matthchr) [inline docs/hugo/content/introduction/multitenant-deployment.md:34]: Nice, this is very clear.

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  AZURE_SUBSCRIPTION_ID: "$AZURE_SUBSCRIPTION_ID"
+  AZURE_TENANT_ID: "$AZURE_TENANT_ID"
+  AZURE_CLIENT_ID: "$AZURE_CLIENT_ID"
+  AZURE_CLIENT_SECRET: "$AZURE_CLIENT_SECRET"
+  AZURE_TARGET_NAMESPACES: "$AZURE_TARGET_NAMESPACES"
+EOF
+```
+
+## Role handling
+The multi-tenant deployment example files have a cluster role granting access to the Azure resource types whichever namespaces they are created in,
```
Comment (matthchr) [inline docs/hugo/content/introduction/multitenant-deployment.md:57]: Phrasing? This part reads weird to me:
"have a cluster role granting access to the Azure resource types whichever namespaces they are created in"

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+Each cluster role binding is named for the specific tenant so they don't collide and can be managed separately.
+This is convenient since it means that you don't need to permit access for Azure resources in each of the target namespaces individually,
+but for some usage scenarios it might be too permissive.
+
+In those cases the `azureserviceoperator-manager-role` should be changed from a `ClusterRole` into `Role`s in each of the target namespaces (where the Azure resources will be created, rather than where the operator pods run),
+and a `RoleBinding` should be created in that namespace linking the `Role` to the service account for the tenant operator that will be managing Azure resources in this target namespace.
+
+## Updgrading
+When upgrading to a newer version of ASO the cluster-wide resources (CRDs, cluster roles) and the webhook operator deployment must be upgraded before upgrading the tenant operators.
+They can be upgraded by applying the new YAML for that type.
```
Comment (matthchr) [inline docs/hugo/content/introduction/multitenant-deployment.md:68]: Is there an ordering requirement between the CRDs and the webhooks, or the CRDs and the cluster roles?

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 1 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:4]: Do you mean something like this?

> The current release version of the ASO deployment YAML installs the operator in single-tenant mode. It:
> * is deployed in the `azureserviceoperator-system` namespace,
> * has a single set of Azure credentials,
> * manages Azure resources in any namespace,
> * handles webhooks fired when any Azure resource is changed.

That feels a bit clunky to me, but if you think it's clearer I'll update it.

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 4 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:16]: Yeah, I agree - I'll reword it to make that clearer.

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 6 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:57]: Yeah, reads like a partial edit... 
Ok, I've updated the wording for that section.

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 7 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:68]: Yes, it should be CRDs then (webhooks and cluster roles). I'll clarify that - what I really mean is that you'd apply the upgraded cluster YAML file first (which will have the changes in the right order), and then the upgraded tenant YAML files.

Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:68]: Ha, "Updgrading"

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 3 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md]: Clarified this (in conjunction with the comment below).

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 1 above)
Comment (matthchr) [reply docs/hugo/content/introduction/multitenant-deployment.md:4]: I did mean that, but agree it is a bit clunky so maybe just leave it as is.

**Review by matthchr — APPROVED** [review]

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
+# Deploying ASO v2 in multi-tenant mode
+
+The current release version of the ASO deployment YAML installs the operator in single-tenant mode:
+it's deployed in the `azureserviceoperator-system` namespace with one set of Azure credentials, and manages resources in any namespace in the cluster.
```
Comment (theunrepentantgeek) [inline docs/hugo/content/introduction/multitenant-deployment.md:4]: Perhaps tweak the wording/formatting to emphasize what's different about single tenant mode?
```suggestion
The current release version of the ASO deployment YAML installs the operator in _single-tenant_ mode:
deployed in the `azureserviceoperator-system` namespace with a _single set of Azure credentials_, and manages resources in _any namespace_ in the cluster.
```

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+That single operator deployment handles webhooks fired when Azure resources are changed.
+
+ASO can also be deployed in a multi-tenant configuration that enables using separate credentials to manage resources in different Kubernetes namespaces.
```
Comment (theunrepentantgeek) [inline docs/hugo/content/introduction/multitenant-deployment.md:7]:
```suggestion
ASO may also be deployed in a _multi-tenant_ configuration, enabling separate credentials for managing resources in different Kubernetes namespaces.
```

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+Running the operator in multi-tenant mode requires one deployment to handle webhooks (since webhook configurations are cluster-level resources) and then a deployment for each tenant, each with its own credentials and set of namespaces that it watches for Azure resources.
+To deploy the operator in multi-tenant mode the release YAML has been split into two parts:
```
Comment (theunrepentantgeek) [inline docs/hugo/content/introduction/multitenant-deployment.md:10]: Minor suggestions for clarity.

```suggestion
Running the operator in multi-tenant mode uses one deployment to handle webhooks (required because webhook configurations are cluster-level resources), and then a separate deployment for each tenant, each with its own credentials and a set of namespaces that it watches for Azure resources.

To deploy the operator in multi-tenant mode the release YAML has been split into two parts:
```

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+1. **Cluster-wide resources**:
+   * Custom resource definitions for the Azure resources.
+
+   * Cluster roles for managing those resources.
+
+   * The `azureserviceoperator-system` namespace containing the deployment and service to handle ASO webhooks.
+     The webhook service is a deployment of the ASO operator image, but configured to run in webhook-only mode.
+     It won't try to reconcile Azure resources with ARM, and so doesn't need Azure credentials.
```
Comment (theunrepentantgeek) [inline docs/hugo/content/introduction/multitenant-deployment.md:19]: Any acronym should be expanded in place and the sentence should make sense.

```suggestion
   * The `azureserviceoperator-system` namespace containing the deployment and service to handle ASO webhooks.
     The webhook service is a deployment of ASO, but configured to run in _webhook-only_ mode.
     It will not try to reconcile Azure resources with ARM, and so does not need any Azure credentials.
```

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+   * A namespace containing the deployment to run the tenant operator, configured for watchers-only mode.
+
+   * The `aso-controller-settings` secret defining the Azure credentials that should be used, and the set of namespaces that this operator will watch for Azure resources.
+
+   * A cluster role binding enabling the per-tenant operator's service account to manage the Azure resources.
+
+## Example files
+Examples of the deployment YAML files are available on the release page for ASO v2 releases from [v2.0.0-alpha.5](https://github.com/Azure/azure-service-operator/releases/tag/v2.0.0-alpha.5).
+The cluster-wide file `multitenant-cluster_v2.0.0-alpha.5.yaml` can be used as-is (the webhook deployment namespace is fixed as `azureserviceoperator-system`),
+but the namespaces and cluster role binding in the per-tenant file `multitenant-tenant_v2.0.0-alpha.5.yaml` will need to be updated from `tenant1` to the desired name for each tenant.
```
Comment (theunrepentantgeek) [inline docs/hugo/content/introduction/multitenant-deployment.md:34]: Trying to make is more apparent that a separate file is needed for each tenant.

```suggestion
The cluster-wide file `multitenant-cluster_v2.0.0-alpha.5.yaml` can be used as-is (the webhook deployment namespace is fixed as `azureserviceoperator-system`),
but the namespaces and cluster role binding in the per-tenant file `multitenant-tenant_v2.0.0-alpha.5.yaml` will need to be customized in each tenant's configuration file from `tenant1` to the desired name for that tenant.
```

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+## Per-tenant configuration
+Create the `aso-controller-settings` secret as described in the [authentication docs](https://azure.github.io/azure-service-operator/introduction/authentication/),
+but create the secret in the tenant namespace and add an extra target namespaces key to it:
```
Comment (theunrepentantgeek) [inline docs/hugo/content/introduction/multitenant-deployment.md:38]: Maybe give an example of what a tenant namespace would look like?

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+## Role handling
+The multi-tenant deployment example files have a single `ClusterRole` that grants access to the Azure resource types,
+and then a binding to that `ClusterRole` for the service account in each tenant-operator namespace.
+Each `ClusterRoleBinding` is named for the specific tenant so they don't collide and can be managed separately.
+
+This is convenient since there's no need to permit access for Azure resources in each of the target namespaces individually,
+but it means that the only thing preventing one tenant operator from reading another's resources is the `AZURE_TARGET_NAMESPACES` setting for each operator.
+For some usage scenarios that might be too permissive.
```
Comment (theunrepentantgeek) [inline docs/hugo/content/introduction/multitenant-deployment.md:69]: Blank line before, so the line stands out more to someone reading the file.

```suggestion

For some usage scenarios that might be too permissive.
```

File: docs/hugo/content/introduction/multitenant-deployment.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+In those cases the `azureserviceoperator-manager-role` should be changed from a `ClusterRole` into `Role`s in each of the target namespaces (where the Azure resources will be created, rather than where the tenant-operator pods run),
+and a `RoleBinding` should be created in that namespace linking the `Role` to the service account for the tenant operator that will be managing Azure resources in this target namespace.
```
Comment (theunrepentantgeek) [inline docs/hugo/content/introduction/multitenant-deployment.md:72]: Would this be clearer with a couple of diagrams showing the differences between scenarios?

**Review by theunrepentantgeek — CHANGES_REQUESTED** [review]

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 11 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:19]: Good call about the acronym, I've added that early in the doc. I'm keeping the distinction that the webhook service is a deployment of the ASO image based on a comment from @matthchr - he felt that calling the webhook service a deployment of the operator was a bit confusing and I agree.

I'm keeping the contractions because I prefer them! 😝

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 12 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:34]: Except I used `YAML file` rather than `configuration file` because I don't want it to be confused with the configuration in the secret below.

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 13 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:38]: Do you mean the contents of the tenant namespace after deployment? Or an example of a tenant name?
I'm assuming the former.

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 15 above)
Comment (babbageclunk) [reply docs/hugo/content/introduction/multitenant-deployment.md:72]: Yeah, good idea - I'll have a go at knocking them up in graphviz.

File: docs/hugo/content/introduction/multitenant-deployment.md (hunk 13 above)
Comment (theunrepentantgeek) [reply docs/hugo/content/introduction/multitenant-deployment.md:40]: I actually meant the later, but showing the former would be good too! I think the diagrams will help here when you embed them inline.

**Review by theunrepentantgeek — APPROVED** [review]
Once those diagrams are inlined, this will be great!
